### PR TITLE
Ddcnl 3750

### PR DESCRIPTION
**Pull request by Sachin300793**
DDCNL 3750
- [ ] Unit tests passing
- [ ] Coverage reached
- [ ] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: app/views/incomeTaxComparison/MainView.scala.html [hunk 1]
```diff
@@ -17,16 +17,17 @@
...
 @import uk.gov.hmrc.tai.model.TaxYear
+@import scala.concurrent.ExecutionContext
 
 @import controllers.auth.AuthenticatedRequest
 
-@this(main: main)
+@this(mainTemplate: MainTemplate)
 
-@(incomeTaxComparisonViewModel: IncomeTaxComparisonViewModel, appConfig: ApplicationConfig)(implicit request: AuthenticatedRequest[_], messages: Messages, templateRenderer: uk.gov.hmrc.renderer.TemplateRenderer, partialRetriever: uk.gov.hmrc.play.partials.FormPartialRetriever)
+@(incomeTaxComparisonViewModel: IncomeTaxComparisonViewModel, appConfig: ApplicationConfig)(implicit request: AuthenticatedRequest[_], messages: Messages, templateRenderer: uk.gov.hmrc.renderer.TemplateRenderer, partialRetriever: uk.gov.hmrc.play.partials.FormPartialRetriever, ec: ExecutionContext)
```
Comment (Sachin300793) [inline app/views/incomeTaxComparison/MainView.scala.html:30]: Can remove FormPartialRetriever

File: app/views/incomes/ConfirmSaveIncomeView.scala.html [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+@import uk.gov.hmrc.play.views.helpers.MoneyPounds
+@import org.joda.time.DateTime
+@import uk.gov.hmrc.play.views.html.layouts.loginStatus
+@import uk.gov.hmrc.tai.util.FormHelper
+@import uk.gov.hmrc.tai.util.TaxYearRangeUtil
+
+@import uk.gov.hmrc.tai.util.MonetaryUtil
+@import uk.gov.hmrc.tai.viewModels.GoogleAnalyticsSettings
+
+@this(mainTemplate: MainTemplate, form: FormWithCSRF)
```
Comment (struong) [inline app/views/incomes/ConfirmSaveIncomeView.scala.html:30]: @Sachin300793 I can't diff this file as its been treated as a new file, can we load it up manually and ensure it looks the same as in Staging?

File: app/views/taxCodeChange/TaxCodeComparisonView.scala.html [hunk 3]
```diff
@@ -14,37 +14,41 @@
...
+@(taxCodeChangeViewModel: TaxCodeChangeViewModel,appConfig: ApplicationConfig)(
+    implicit request: Request[_],
+    user: controllers.auth.AuthedUser,
+    messages: Messages,
+    templateRenderer: uk.gov.hmrc.renderer.TemplateRenderer,
+    ec:  scala.concurrent.ExecutionContext)
 
 @gaEventActionString = @{"Your tax code change"}
-@gaEventLabelString = @{Messages("taxCode.change.yourTaxCodeChanged.checkYourTaxButton")}
+@gaEventLabelString = @{messages("taxCode.change.yourTaxCodeChanged.checkYourTaxButton")}
```
Comment (struong) [inline app/views/taxCodeChange/TaxCodeComparisonView.scala.html:39]: @Sachin300793 why has messages changes throughout this file randomly? They should all be lower case messages here as this is the one we are implicitly passing in

File: app/views/taxCodeChange/TaxCodeComparisonView.scala.html [hunk 4]
```diff
@@ -74,7 +78,7 @@
 
 @taxCodeRecordPod(taxCodeRecord: Option[TaxCodeRecord], identifier: String) = {
     @{taxCodeRecord match {
-        case Some(taxCodeRecord) => formattedTaxCodeForComparison(taxCodeRecord, identifier, taxCodeChangeViewModel.scottishTaxRateBands, appConfig)
+        case Some(taxCodeRecord) => FormattedTaxCodeForComparison(taxCodeRecord, identifier, taxCodeChangeViewModel.scottishTaxRateBands, appConfig)
```
Comment (struong) [inline app/views/taxCodeChange/TaxCodeComparisonView.scala.html:81]: @Sachin300793 just checking, should this have a capital F?

File: test/controllers/TaxCodeChangeControllerSpec.scala [hunk 5]
```diff
@@ -152,9 +152,7 @@ class TaxCodeChangeControllerSpec extends BaseSpec with ControllerViewTestHelper
         mcc,
         taxCodeComparisonView,
         yourTaxFreeAmountView,
-        whatHappensNextView,
-        partialRetriever,
-        templateRenderer
+        whatHappensNextView
```
Comment (struong) [inline test/controllers/TaxCodeChangeControllerSpec.scala:155]: don't we still need `templateRenderer`?

File: test/uk/gov/hmrc/tai/connectors/TaxCodeChangeConnectorSpec.scala [hunk 6]
```diff
@@ -160,31 +160,31 @@ class TaxCodeChangeConnectorSpec extends BaseSpec {
...
+      val expectedResult = List(taxCodeRecord, taxCodeRecord2)
 
       when(httpHandler.getFromApi(Matchers.eq(latestTaxCodeRecordUrl))(any())).thenReturn(Future.successful(json))
 
       val result = Await.result(sut.lastTaxCodeRecords(nino, TaxYear().prev), 5 seconds)
       result mustEqual TaiSuccessResponseWithPayload(expectedResult)
     }
 
-    "return a empty sequence when the api returns no records" in {
+    "return a empty Listuence when the api returns no records" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/TaxCodeChangeConnectorSpec.scala:174]: Listuence?

File: test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala [hunk 7]
```diff
@@ -30,42 +30,42 @@ class TaxCodeChangeSpec extends PlaySpec {
...
 
-      "throw a JsError given an empty Seq of TaxCodeRecords" in {
+      "throw a JsError given an empty List of TaxCodeRecords" in {
         an[JsResultException] should be thrownBy emptyTaxCodeRecordsJson.as[TaxCodeChange]
       }
     }
 
     "mostRecentTaxCodeChangeDate" should {
-      "return the latest tax code change date from a sequence of tax code records" in {
+      "return the latest tax code change date from a Listuence of tax code records" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala:46]: Listuence?

File: test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala [hunk 8]
```diff
@@ -30,42 +30,42 @@ class TaxCodeChangeSpec extends PlaySpec {
...
-          TaxCodeChange(Seq(previousTaxCodeRecord1, fullYearTaxCode), Seq(currentTaxCodeRecord1, fullYearTaxCode))
+          TaxCodeChange(List(previousTaxCodeRecord1, fullYearTaxCode), List(currentTaxCodeRecord1, fullYearTaxCode))
 
         model.mostRecentTaxCodeChangeDate mustEqual startDate.plusMonths(1).plusDays(1)
       }
     }
 
     "mostRecentPreviousTaxCodeChangeDate" should {
-      "return the latest tax code change date from a sequence of tax code records" in {
+      "return the latest tax code change date from a Listuence of tax code records" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala:55]: Listuence? Please be more cautious when mass renaming, a better selection would have been to match on case + word and on `Seq(` -> `List(`

File: test/uk/gov/hmrc/tai/util/HelpFormHelperSpec.scala [hunk 9]
```diff
@@ -20,12 +20,12 @@ import builders.RequestBuilder
...
 import org.mockito.Mockito._
 import play.twirl.api.Html
-import uk.gov.hmrc.play.partials.FormPartialRetriever
+import uk.gov.hmrc.play.partials.PartialRetriever
 import utils.BaseSpec
 
 class HelpFormHelperSpec extends BaseSpec {
 
-  val mockPartialRetriever: FormPartialRetriever = mock[FormPartialRetriever]
+  val mockPartialRetriever: PartialRetriever = mock[PartialRetriever]
```
Comment (struong) [inline test/uk/gov/hmrc/tai/util/HelpFormHelperSpec.scala:28]: @Sachin300793 this is my mistake, this should still be `FormPartialRetriever`

Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/util/HelpFormHelperSpec.scala:28]: Done

File: test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala (hunk 8 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala:55]: Will take care , done

File: test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala (hunk 7 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/model/domain/TaxCodeChangeSpec.scala:46]: Corrected to List , done

File: test/uk/gov/hmrc/tai/connectors/TaxCodeChangeConnectorSpec.scala (hunk 6 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/TaxCodeChangeConnectorSpec.scala:174]: Corrected to List , done

File: app/uk/gov/hmrc/tai/util/HelpFormHelper.scala [hunk 10]
```diff
@@ -19,14 +19,17 @@ package uk.gov.hmrc.tai.util
...
-import uk.gov.hmrc.play.partials.FormPartialRetriever
+import uk.gov.hmrc.play.partials.PartialRetriever
 import uk.gov.hmrc.tai.config.ApplicationConfig
 
+import scala.concurrent.ExecutionContext
+
 object HelpFormHelper {
 
-  def replaceMessage(partialRetriever: FormPartialRetriever, applicationConfig: ApplicationConfig)(
+  def replaceMessage(partialRetriever: PartialRetriever, applicationConfig: ApplicationConfig)(
```
[the remark this answers is not in the document]
Comment (Sachin300793) [reply app/uk/gov/hmrc/tai/util/HelpFormHelper.scala:29]: Done

File: app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala [hunk 11]
```diff
@@ -28,22 +26,21 @@ object TaxYearObjectBinder {
...
-        lazy val appConfig = Play.current.injector.instanceOf[ApplicationConfig]
-
         val taxYearRequired: TaxYear = TaxYear(value)
 
         val currentYear: Int = TaxYear().year
-
         val latestSupportedTaxYear = TaxYear().next
-        val earliestSupportedTaxYear = TaxYear(currentYear - appConfig.numberOfPreviousYearsToShow)
+
+        val numberOfPreviousYearsToShow = 5
```
Comment (Christian-QA) [inline app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala:34]: Isn't this a magic number? Shouldn't this be stored in the config file to make it easier to find if it's changed?

File: app/views/incomes/ConfirmSaveIncomeView.scala.html (hunk 2 above)
Comment (Sachin300793) [reply app/views/incomes/ConfirmSaveIncomeView.scala.html:30]: Its name has been changed from confirm_save_Income.scala.html to ConfirmSaveIncomeView.scala.html . We removed partialRetriever: uk.gov.hmrc.play.partials.FormPartialRetriever and changed main to MainTemplate

File: app/uk/gov/hmrc/tai/config/ApplicationConfig.scala [hunk 12]
```diff
@@ -140,7 +140,8 @@ trait FeatureTogglesConfig { self: ApplicationConfig =>
   lazy val accessibilityStatementToggle: Boolean =
     getOptional[Boolean]("accessibility-statement.toggle").getOrElse(false)
   lazy val isTaiCy3Enabled: Boolean = getOptional[Boolean]("tai.cy3.enabled").getOrElse(false)
-  lazy val numberOfPreviousYearsToShow: Int = getOptional[Int]("tai.numberOfPreviousYearsToShow").getOrElse(5)
+
+  val numberOfPreviousYearsToShow: Int = 5 // Always 5 in all configs
```
Comment (Christian-QA) [inline app/uk/gov/hmrc/tai/config/ApplicationConfig.scala:144]: Answers my question from above

File: test/controllers/TaxCodeChangeControllerSpec.scala (hunk 5 above)
Comment (Sachin300793) [reply test/controllers/TaxCodeChangeControllerSpec.scala:155]: Moved templateRenderer to implicit block

File: app/views/taxCodeChange/TaxCodeComparisonView.scala.html (hunk 4 above)
Comment (Sachin300793) [reply app/views/taxCodeChange/TaxCodeComparisonView.scala.html:81]: Changed the file name from formattedTaxCodeForComparison to FormattedTaxCodeForComparison

File: app/views/unauthenticated_main.scala.html [hunk 13]
```diff
@@ -14,26 +14,17 @@
...
-    trackingConsentSnippet: TrackingConsentSnippet,
-    article: Article,
-    sidebar: Sidebar,
-    attorney_banner: AttorneyBanner
+        appConfig: ApplicationConfig,
+        langUtils: LanguageUtils,
+        trackingConsentSnippet: TrackingConsentSnippet,
+        article: Article,
+        sidebar: Sidebar,
+        attorney_banner: AttorneyBanner
```
Comment (Christian-QA) [inline app/views/unauthenticated_main.scala.html:27]: Can we camel case attorney_banner?

File: app/views/taxCodeChange/TaxCodeComparisonView.scala.html (hunk 3 above)
Comment (Sachin300793) [reply app/views/taxCodeChange/TaxCodeComparisonView.scala.html:39]: Made changes to use lower case messages , done

File: app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala (hunk 11 above)
Comment (Christian-QA) [reply app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala:34]: If you check the next comment, there's already a value stored in the ApplicationConfig that you can use

File: app/controllers/EstimatedIncomeTaxController.scala [hunk 14]
```diff
@@ -45,12 +45,13 @@ class EstimatedIncomeTaxController @Inject()(
   complexEstimatedIncomeTax: ComplexEstimatedIncomeTaxView,
   simpleEstimatedIncomeTax: SimpleEstimatedIncomeTaxView,
   zeroTaxEstimatedIncomeTax: ZeroTaxEstimatedIncomeTaxView,
-  implicit val partialRetriever: FormPartialRetriever,
   implicit val templateRenderer: TemplateRenderer,
   mcc: MessagesControllerComponents,
   errorPagesHandler: ErrorPagesHandler)(implicit ec: ExecutionContext)
     extends TaiBaseController(mcc) {
 
+  // TODO -  /paye-income-tax-estimate  check that if partials work for it
```
[the remark this answers is not in the document]
Comment (Sachin300793) [reply app/controllers/EstimatedIncomeTaxController.scala:53]: Done

File: app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala (hunk 11 above)
Comment (Sachin300793) [reply app/uk/gov/hmrc/tai/binders/TaxYearObjectBinder.scala:34]: Can't inject appConfig so hard coded here

File: app/views/unauthenticated_main.scala.html (hunk 13 above)
Comment (Sachin300793) [reply app/views/unauthenticated_main.scala.html:27]: Done

File: app/uk/gov/hmrc/tai/config/ApplicationConfig.scala (hunk 12 above)
Comment (Sachin300793) [reply app/uk/gov/hmrc/tai/config/ApplicationConfig.scala:144]: Can't inject appConfig so hard coded here

**Review by struong — APPROVED** [review]

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 15]
```diff
@@ -16,27 +16,40 @@
...
-class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
+class HttpHandlerSpec
+    extends WordSpec with GuiceOneAppPerSuite with MustMatchers with WireMockHelper with ScalaFutures
+    with IntegrationPatience with Injecting {
 
-  override def beforeEach: Unit =
-    Mockito.reset(http)
+  val generatedNino = new Generator().nextNino
+
+  val generatedSaUtr = new Generator().nextAtedUtr
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:41]: Do we need this? Not sure how a SaUtr is linked to an AtedUtr here? TAI also does not use SAUtrs anywhere.

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 16]
```diff
@@ -16,27 +16,40 @@
...
+
+  lazy val messages = inject[Messages]
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:43]: Is this used?

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 17]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
     }
 
-    "result in a Locked exception" when {
-      "when a Locked response is received from the http get call" in {
-        when(http.GET[HttpResponse](any())(any(), any(), any())).thenReturn(Future.successful(LockedHttpResponse))
-        val responseFuture = sut.getFromApiV2("")
-        val ex = the[LockedException] thrownBy Await.result(responseFuture, 5 seconds)
-        ex.message mustBe "\"locked\""
-      }
+    "should return a LockedException when LOCKED response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:113]: this is throwing a `Upstream4xxResponse` not a `LockedException`

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 18]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+      val ex = the[UnauthorizedException] thrownBy Await.result(responseFuture, 5 seconds)
+      ex.message must include("unauthorized")
+
     }
 
-    "result in an UnauthorisedException" when {
-      "when an Unauthorised response is received from the http get call" in {
-        val unauthorisedResponse =
-          HttpResponse(UNAUTHORIZED, Some(JsString("unauthorised response")), Map("ETag" -> Seq("34")))
+    "should return a HttpException when unknown response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:137]: This is not quite true 4xx statuses will throw a Upstream4xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 19]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+      val result = Await.result(httpHandler.putToApi[DateRequest](testUrl, DateRequest(LocalDate.now())), 5.seconds)
 
       result.status mustBe OK
 
     }
 
-    "return Not Found exception" in {
-      when(http.PUT[DateRequest, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(NOT_FOUND)))
+    "should return a NotFoundException when NOT_FOUND response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:165]: This code is not throwing an exception?

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 20]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+        .result(httpHandler.putToApi[DateRequest](testUrl, DateRequest(LocalDate.now())), 5.seconds)
 
       result.responseCode mustBe NOT_FOUND
+
     }
 
-    "return Internal Server exception" in {
-      when(http.PUT[DateRequest, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(INTERNAL_SERVER_ERROR)))
+    "should return a InternalServerException when INTERNAL_SERVER_ERROR response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:178]: This code is throwing a Upstream5xxResponse not returning a HTTP code

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 21]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+
+      ex.message must include("internal server exception")
 
-      result.responseCode mustBe INTERNAL_SERVER_ERROR
     }
 
-    "return Bad Request exception" in {
-      when(http.PUT[DateRequest, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(BAD_REQUEST)))
+    "should return a BadRequestException when BAD_REQUEST response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:191]: This code is not throwing an exception, it is returning a HTTP status code

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 22]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+        .result(httpHandler.putToApi[DateRequest](testUrl, DateRequest(LocalDate.now())), 5.seconds)
 
       result.responseCode mustBe BAD_REQUEST
+
     }
 
-    "return Http exception" in {
-      when(http.PUT[DateRequest, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(GATEWAY_TIMEOUT)))
+    "should return a HttpException when unknown response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:204]: This code is not throwing a HttpException, its treating 4xx statuses as a Upstream4xxResponse exception

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 23]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
-  "postToApi" should {
+  "postToApi" must {
     val userInput = "userInput"
 
-    "return json which is coming from http post call" in {
+    List(
+      CREATED,
+      OK
+    ).foreach { httpStatus =>
+      s"return json which is coming from http post call for $httpStatus response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:225]: This code is not returning json

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 24]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
-    "return Http exception" when {
-      "http response is NOT_FOUND" in {
-        when(http.POST[String, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-          .thenReturn(Future.successful(HttpResponse(NOT_FOUND)))
+    List(
+      GATEWAY_TIMEOUT,
+      INTERNAL_SERVER_ERROR,
+      SERVICE_UNAVAILABLE
+    ).foreach { httpStatus =>
+      s"return Http exception for $httpStatus response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:243]: This is not quite true, its returning a Upstream5xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 25]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+        ex.message must include("error response")
 
-        result.responseCode mustBe NOT_FOUND
       }
+    }
 
-      "http response is GATEWAY_TIMEOUT" in {
-        when(http.POST[String, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-          .thenReturn(Future.successful(HttpResponse(GATEWAY_TIMEOUT)))
+    "return Http exception for BAD_REQUEST response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:256]: This is not throwing an exception

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 26]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+        post(anyUrl())
+          .willReturn(aResponse().withStatus(BAD_REQUEST)))
+      val result = the[BadRequestException] thrownBy Await
+        .result(httpHandler.postToApi[String](testUrl, userInput), 5 seconds)
 
-        val result = the[HttpException] thrownBy Await.result(sut.postToApi[String](mockUrl, userInput), 5 seconds)
+      result.responseCode mustBe BAD_REQUEST
+    }
+
+    "return Http exception for NOT_FOUND response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:266]: This is not throwing an exception

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 27]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
-          .thenReturn(Future.successful(HttpResponse(OK)))
-        val result = Await.result(sut.deleteFromApi(mockUrl), 5 seconds)
-        result.status mustBe OK
-      }
+    List(
+      ACCEPTED,
+      OK,
+      NO_CONTENT
+    ).foreach { httpStatus =>
+      s"return json which is coming from http post call for $httpStatus response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:289]: This is not returning json

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 28]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
-        when(http.DELETE[HttpResponse](any(), any())(any(), any(), any()))
-          .thenReturn(Future.successful(HttpResponse(GATEWAY_TIMEOUT)))
 
-        val result = the[HttpException] thrownBy Await.result(sut.deleteFromApi(mockUrl), 5 seconds)
+    List(
+      GATEWAY_TIMEOUT,
+      INTERNAL_SERVER_ERROR,
+      SERVICE_UNAVAILABLE
+    ).foreach { httpStatus =>
+      s"return Http exception for $httpStatus response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:306]: This is not quite true, it is throwing a Upstream5xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 29]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+        val ex = the[Upstream5xxResponse] thrownBy Await.result(responseFuture, 5 seconds)
+        ex.message must include("error response")
 
-        result.responseCode mustBe GATEWAY_TIMEOUT
       }
     }
-  }
 
-  private val mockUrl = "mockUrl"
+    "return Http exception for BAD_REQUEST response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:319]: This is throwing a BadRequestException, we should be as specific as we can

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala [hunk 30]
```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
...
+      server.stubFor(
+        delete(anyUrl())
+          .willReturn(aResponse().withStatus(BAD_REQUEST).withBody("bad request")))
+      val responseFuture = httpHandler.deleteFromApi(testUrl)
+      val ex = the[BadRequestException] thrownBy Await.result(responseFuture, 5 seconds)
+      ex.message must include("bad request")
+    }
 
-  private val responseBodyObject = ResponseObject("Name", 24)
+    "return Http exception for NOT_FOUND response" in {
```
Comment (struong) [inline test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:328]: Same as above but a NotFoundException

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 15 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:41]: Removed them , done

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 16 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:43]: Removed them , done

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 17 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:113]: Changed test name to Upstream4xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 18 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:137]: Changed its name to Upstream4xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 30 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:328]: changed the name to NotFoundException

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 29 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:319]: changed the name to BadRequestException

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 28 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:306]: Chnaged the name to Upstream5xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 27 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:289]: changed name to  s"return $httpStatus status from http post call for $httpStatus response"

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 20 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:178]: Changed name to Upstream5xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 22 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:204]: Changed name to Upstream4xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 23 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:225]: chnaged it name to  s"return $httpStatus response status for $httpStatus response"

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 24 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:243]: Changed name to Upstream5xxResponse

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 19 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:165]: We are checking responseCode of the response here

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 21 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:191]: Changing name to should return a BAD_REQUEST response code when BAD_REQUEST response

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 25 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:256]: changed name to should return a BAD_REQUEST response code when BAD_REQUEST response

File: test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala (hunk 26 above)
Comment (Sachin300793) [reply test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala:266]: changed name to should return a NOT_FOUND response code when NOT_FOUND response

**Review by struong — APPROVED** [review]

**Review by struong — APPROVED** [review]

**Review by brendanstraughan — APPROVED** [review]

**Review by struong — APPROVED** [review]

**Pull request merged**